src/ML-METATENSOR/metatensor_system.h

```diff
@@ -0,0 +1,135 @@
+/* -*- c++ -*- ----------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#ifndef LMP_METATENSOR_SYSTEM_H
+#define LMP_METATENSOR_SYSTEM_H
+
+#include <vector>
+#include <array>
+#include <unordered_set>
+
+#include "pointers.h"
+#include "pair.h"
+#include "neigh_list.h"
+
+#include <metatensor/torch/atomistic.hpp>
+
+
+namespace LAMMPS_NS {
+
+struct MetatensorSystemOptions {
+    // Mapping from LAMMPS types to metatensor types
+    const int32_t* types_mapping;
+    // interaction range of the model, in LAMMPS units
+    double interaction_range;
+    // should we run extra checks on the neighbor lists?
+    bool check_consistency;
+};
+
+// data for metatensor neighbors lists
+struct MetatensorNeighborsData {
+    // single neighbors sample containing [i, j, S_a, S_b, S_c]
+    using sample_t = std::array<int32_t, 5>;
+
+    struct SampleHasher {
+        static void hash_combine(std::size_t& seed, const int32_t& v) {
+            seed ^= std::hash<int32_t>()(v) + 0x9e3779b9 + (seed<<6) + (seed>>2);
+        }
+
+        size_t operator()(const sample_t& s) const {
+            size_t hash = 0;
+            hash_combine(hash, s[0]);
+            hash_combine(hash, s[1]);
+            hash_combine(hash, s[2]);
+            hash_combine(hash, s[3]);
+            hash_combine(hash, s[4]);
+            return hash;
+        }
+    };
```

Comment:
Is this necessary? What do other packages do? I assume that they just have atoms from 0 to n_atoms-1 and then they remove duplicates according to their atom index

Reply:
This is de-duplicating full neighbor list samples (i.e. `i, j, S_a, S_b, S_c` tuples) by storing them in a `std::unordered_set`

The unordered_set will first compute the hash, and if it does not match an existing entry then insert the new one. If the hash matches, then unordered_set will actually compare all values in the sample for equality.

Reply:
most other ML-XXX packages either use LAMMPS NL as-is, or re-compute a new NL.

Reply:
Ok, fair, and what can give rise to such a duplication?

Reply:
if you have a very large cutoff and a small cell, LAMMPS will create multiple copies/ghost versions of the atoms in the main simulation box. This means you can have a pair between atom 124 and 745 which in metatensor world is a pair (3, 5, 1, 2, -1); and then another pair for 912 and 834 which is also (3, 5, 1, 2, -1); if the atoms in both pairs are shifted the same number of times, i.e. 912 is 124 + (1, 0, 1) and 834 is 745 + (1, 0, 1).